subprojects/core/src/main/java/org/gradle/initialization/ParallelismBuildOptionFactory.java

```diff
@@ -48,33 +49,38 @@ public ParallelOption() {
         }
 
         @Override
-        public void applyTo(boolean value, ParallelismConfiguration settings) {
+        public void applyTo(boolean value, ParallelismConfiguration settings, Origin origin) {
             settings.setParallelProjectExecutionEnabled(value);
         }
     }
 
     public static class MaxWorkersOption extends StringBuildOption<ParallelismConfiguration> {
         public static final String GRADLE_PROPERTY = "org.gradle.workers.max";
+        public static final String LONG_OPTION = "max-workers";
 
         public MaxWorkersOption() {
             super(GRADLE_PROPERTY, CommandLineOptionConfiguration.create("max-workers", "Configure the number of concurrent workers Gradle is allowed to use.").incubating());
         }
 
         @Override
-        public void applyTo(String value, ParallelismConfiguration settings) {
+        public void applyTo(String value, ParallelismConfiguration settings, Origin origin) {
             try {
                 int workerCount = Integer.parseInt(value);
                 if (workerCount < 1) {
-                    invalidMaxWorkersSwitchValue(value);
+                    handleInvalidMaxWorkersSwitchValue(value, origin);
                 }
                 settings.setMaxWorkerCount(workerCount);
             } catch (NumberFormatException e) {
-                invalidMaxWorkersSwitchValue(value);
+                handleInvalidMaxWorkersSwitchValue(value, origin);
             }
         }
 
-        private void invalidMaxWorkersSwitchValue(String value) {
-            throw new IllegalArgumentException(String.format("Argument value '%s' given for system property %s or --%s option is invalid (must be a positive, non-zero, integer)", value, gradleProperty, commandLineOptionConfiguration.getLongOption()));
+        private void handleInvalidMaxWorkersSwitchValue(String value, Origin origin) {
+            switch (origin) {
+                case GRADLE_PROPERTY: throw new IllegalArgumentException(String.format(String.format("Value '%s' given for %s system property is invalid (must be a positive, non-zero, integer)", value, gradleProperty)));
```

Comment:
This could probably be extracted (because other places might have validation too and would then have similar switch statements). Instead the `Origin` class could have a method to throw the appropriate exception (taking an actual and expected value).

Reply:
done